src/app/model/ChessEngine.ts

```diff
@@ -1,3 +1,57 @@
 import { ChessModel } from './ChessModel';
+import { Side, Square } from './Types';
+import { Piece } from './Piece';
+import { Player } from './Player';
+import { ChessBoard } from './ChessBoard';
+import { PieceWasMoved } from './PieceWasMoved';
+import { PieceWasCaptured } from './PieceWasCaptured';
+import { isDefined } from './HelperFunctions';
 
-export class ChessEngine implements ChessModel {}
+export class ChessEngine implements ChessModel {
+  private currentSide: Side = Side.BLACK;
+
+  constructor(readonly board: ChessBoard) {}
+
+  move(byPlayer: Player, squareFrom: Square, squareTo: Square): (PieceWasMoved | PieceWasCaptured)[] {
+    const chosenPiece = this.board.onPositionPiece(squareFrom);
+    if (!chosenPiece) {
+      throw new Error('There is no piece on this square.');
+    }
+    if (byPlayer.side == this.currentSide) {
+      throw new Error('Player can not move twice in a row.');
+    }
+    if (byPlayer.side !== chosenPiece.side) {
+      throw new Error('Player can not move other players pieces.');
+    }
+    const possibleMoves = chosenPiece.possibleMoves(squareFrom, this.board);
+    if (!this.canMoveOnSquare(possibleMoves, squareTo)) {
```

Comment:
Nawet chyba lepiej do tej metodki dac caly chosenPiece i niech tam w srodku wyciaga :)

Reply:
to  w takim razie muszę przekazać do tej metody squareFrom, bo to na tej podstawie okreśła się piece i jego ruchy :)